include/kllvm/binary/deserializer.h

```diff
@@ -203,6 +205,146 @@ class proof_trace_file_buffer : public proof_trace_buffer {
   }
 };
 
+class proof_trace_ringbuffer : public proof_trace_buffer {
+private:
+  shm_ringbuffer *shm_buffer_;
+  sem_t *data_avail_;
+  sem_t *space_avail_;
+  std::deque<uint8_t> peek_data_;
+  bool peek_eof_{false};
+  bool read_eof_{false};
+
+  bool read(uint8_t *ptr, size_t len = 1) {
+    for (size_t i = 0; i < len; i++) {
+      if (!peek_data_.empty()) {
+        ptr[i] = peek_data_.front();
+        peek_data_.pop_front();
+        continue;
+      }
+
+      if (peek_eof_) {
+        read_eof_ = true;
+        return false;
+      }
+
+      if (read_eof_) {
+        return false;
+      }
+
+      while (int wait_status = sem_trywait(data_avail_)) {
+        assert(wait_status == -1 && errno == EAGAIN);
+        if (shm_buffer_->eof()) {
+          read_eof_ = true;
+          return false;
+        }
+      }
+      shm_buffer_->get(&ptr[i]);
+      sem_post(space_avail_);
+    }
+
+    return true;
+  }
+
+  bool peek(uint8_t *ptr, size_t len = 1) {
+    for (size_t i = 0; i < len; i++) {
+      if (i < peek_data_.size()) {
+        ptr[i] = peek_data_[i];
+        continue;
+      }
+
+      if (peek_eof_) {
+        return false;
+      }
+
+      if (read_eof_) {
+        return false;
+      }
+
+      while (int wait_status = sem_trywait(data_avail_)) {
+        assert(wait_status == -1 && errno == EAGAIN);
+        if (shm_buffer_->eof()) {
+          peek_eof_ = true;
+          return false;
+        }
+      }
+      shm_buffer_->get(&ptr[i]);
+      sem_post(space_avail_);
+      peek_data_.push_back(ptr[i]);
+    }
+
+    return true;
+  }
+
+public:
+  proof_trace_ringbuffer(
+      shm_ringbuffer *buffer, sem_t *data_avail, sem_t *space_avail)
+      : shm_buffer_(buffer)
+      , data_avail_(data_avail)
+      , space_avail_(space_avail) { }
```

Comment:
One more minor nitpick: the destructor of the `shm_ringbuffer` object should be called explicitly somewhere. Currently this isn't an obvious issue because there are no resources being managed by the buffer (just plain member variables), but if this changes in the future you'll have a subtle issue from the missing destructor call.

The way I'd suggest organising this is to have `proof_trace_ringbuffer` take a `void *` argument and perform the placement new in its constructor, then call the destructor of the created buffer in its destructor. An example of what that might look like: https://godbolt.org/z/dfb5cTYns

Otherwise looks good; feel free to merge once this and my other minor nits are resolved.

Reply:
I have addressed this and merged all your suggestions. I will merge the PR when CI passes.